improve handling of NEWS navigation

kevinushey opened the pull request:
This PR improves our ability to navigate to the CRAN NEWS file associated with a currently installed package.

Effectively, RStudio now tries multiple sets of links until it discovers one that does not 404; after discovering such a link it opens that link in the browser.

Note that with this change, we always show a NEWS link icon, but provide an error message to the user if we failed to figure out what the correct NEWS link is.

Closes #3082.

jmcphers commented on src/cpp/session/modules/SessionPackages.R:
```diff
@@ -430,44 +433,93 @@ if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
+   else if (length(repos))
+      repos[[1]]
+   else
+      "https://cran.rstudio.com"
```
This should check `UserSettings` for the default repo (as in the previous PR)

jmcphers commented on src/cpp/session/modules/SessionPackages.R:
```diff
@@ -430,44 +433,93 @@ if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
+      
+      url <- file.path(prefix, packageName, candidate)
+      
+      status <- .rs.withTimeLimit(5, {
```
If there's a connectivity problem then this could take a long time as we timeout for each URL (and it'll just look hung to the user). Maybe we should abort on network errors?

jmcphers replied:
Also, should we use a network timeout parameter (e.g. `options(timeout)`) instead of having R kill the request from the top level?

kevinushey replied:
Good idea -- that's a much better solution here.

jmcphers commented on src/cpp/session/modules/SessionPackages.R:
```diff
@@ -430,44 +433,93 @@ if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
+      url <- file.path(prefix, packageName, candidate)
+      
+      status <- .rs.withTimeLimit(5, {
+         .rs.tryCatch(download.file(url, destfile = tempfile(), quiet = TRUE, mode = "wb"))
```
I think we can use `HEAD` to test for the URL's existence without downloading the whole thing, although possibly not worth doing if R doesn't make it easy.

kevinushey replied:
Given that the NEWS files tend to be only a few kilobytes in size I think it should be okay to just attempt to download the file directly. (You're correct that there isn't a simple, dependency-less way of making a `HEAD` request from R)

I thought about whether it'd be worth exposing some of our C++ machinery for doing this, but using `download.file()` also ensures we respect all of the other download options set in the R session (`download.file.method`, `download.file.extra`, any options or environment variables affecting `curl` or `wget`, and so on)

jmcphers commented on src/cpp/session/modules/SessionPackages.R:
```diff
@@ -430,44 +433,93 @@ if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
+      }
+   }
+   
+   return(NULL)
```
In my view we should return our best guess at the URL even if we failed to get one. That way whatever error was observed in `download.file` (which the user can't see) will show up in their browser. What do you think?

kevinushey commented:
I think I've addressed all the comments. I also did some munging of `download.file.extra` for `curl`, since it doesn't fail on 404 by default. (congratulations, you successfully retrieved the 404 web page!)

I thought about whether we could just force the option for `download.file.method` but I figure that if the user has explicitly configured `curl` it's likely to support some proxy configuration and so we should do the extra work to pass the correct command line arguments as we need them.
